## Final Report

In [1]:
import pandas as pd
import numpy as np
import wrangle as w

import warnings
warnings.filterwarnings("ignore")

pd.set_option('display.max_rows', 1000)

In [2]:
#get the data
df = w.acquire_anonymized_curriculum_access_data()

In [3]:
#look at the data
df

date      time  \
0       2018-01-26  09:55:03   
1       2018-01-26  09:56:02   
2       2018-01-26  09:56:05   
3       2018-01-26  09:56:06   
4       2018-01-26  09:56:24   
...            ...       ...   
900218  2021-04-21  16:41:51   
900219  2021-04-21  16:42:02   
900220  2021-04-21  16:42:09   
900221  2021-04-21  16:44:37   
900222  2021-04-21  16:44:39   

                                                    page  user_id  cohort_id  \
0                                                      /        1        8.0   
1                                                java-ii        1        8.0   
2                    java-ii/object-oriented-programming        1        8.0   
3                     slides/object_oriented_programming        1        8.0   
4                              javascript-i/conditionals        2       22.0   
...                                                  ...      ...        ...   
900218                              jquery/personal-site       64       28.0   
900219                                 jquery/mapbox-api       64       28.0   
900220                           jquery/ajax/weather-map       64       28.0   
900221  anomaly-detection/discrete-probabilistic-methods      744       28.0   
900222                                 jquery/mapbox-api       64       28.0   

                   ip     name  start_date    end_date  program_id  
0        97.105.19.61  Hampton  2015-09-22  2016-02-06         1.0  
1        97.105.19.61  Hampton  2015-09-22  2016-02-06         1.0  
2        97.105.19.61  Hampton  2015-09-22  2016-02-06         1.0  
3        97.105.19.61  Hampton  2015-09-22  2016-02-06         1.0  
4        97.105.19.61    Teddy  2018-01-08  2018-05-17         2.0  
...               ...      ...         ...         ...         ...  
900218  71.150.217.33    Staff  2014-02-04  2014-02-04         2.0  
900219  71.150.217.33    Staff  2014-02-04  2014-02-04         2.0  
900220  71.150.217.33    Staff  2014-02-04  2014-02-04         2.0  
900221  24.160.137.86    Staff  2014-02-04  2014-02-04         2.0  
900222  71.150.217.33    Staff  2014-02-04  2014-02-04         2.0  

[900223 rows x 10 columns]

In [4]:
#look at the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 900223 entries, 0 to 900222
Data columns (total 10 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   date        900223 non-null  object 
 1   time        900223 non-null  object 
 2   page        900222 non-null  object 
 3   user_id     900223 non-null  int64  
 4   cohort_id   847330 non-null  float64
 5   ip          900223 non-null  object 
 6   name        847330 non-null  object 
 7   start_date  847330 non-null  object 
 8   end_date    847330 non-null  object 
 9   program_id  847330 non-null  float64
dtypes: float64(2), int64(1), object(7)
memory usage: 68.7+ MB


In [5]:
# clean the data
df = w.clean_the_data(df)

## 1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?

**Findings:**
- Data Science accessed classification/overview 1785 times.
- Front End accessed content/html-css 2 times.
- Java accessed javascript-i 17457 times.
- PHP accessed javascript-i 736 times.

In [8]:
#Make df for each program
ds=df[df.program_id=='Data Science']
fe=df[df.program_id=='Front End']
jv=df[df.program_id=='Java']
php=df[df.program_id=='PHP']

In [16]:
#Look at what page was most viewed for data science
ds.groupby(['page'])['page'].count().sort_values(ascending=False)

page
/                                           8358
search/search_index.json                    2203
classification/overview                     1785
1-fundamentals/modern-data-scientist.jpg    1655
1-fundamentals/AI-ML-DL-timeline.jpg        1651
                                            ... 
cli/4-navigating-the-filesystem                1
regression/project/null                        1
creating-charts                                1
curie-statistics-assessment                    1
%20https://github.com/RaulCPena                1
Name: page, Length: 682, dtype: int64

In [17]:
#Look at what page was most viewed for front end
fe.groupby(['page'])['page'].count().sort_values(ascending=False)

page
content/html-css                               2
/                                              1
content/html-css/gitbook/images/favicon.ico    1
content/html-css/introduction.html             1
Name: page, dtype: int64

In [18]:
#Look at what page was viewed the most for java
jv.groupby(['page'])['page'].count().sort_values(ascending=False)

page
/                                                                      35814
javascript-i                                                           17457
toc                                                                    17428
search/search_index.json                                               15212
java-iii                                                               12683
                                                                       ...  
content/mysql/relationships/sub-queries.html                               1
content/mysql/sample-database/gitbook/images/favicon.ico                   1
content/php_iii/classes-and-objects-i/constructors-destructors.html        1
content/php_iv/classes-and-objects-ii/late-static-binding.html             1
wp-login                                                                   1
Name: page, Length: 1913, dtype: int64

In [19]:
#Look at what page was viewed the most for php
php.groupby(['page'])['page'].count().sort_values(ascending=False)

page
/                      1681
index.html             1011
javascript-i            736
html-css                542
spring                  501
                       ... 
interview-questions       1
13.5_Tableau              1
interview                 1
html/forms                1
4.5_imports               1
Name: page, Length: 710, dtype: int64

## 2. Is there a cohort that referred to a lesson significantly more than other cohorts seemed to gloss over?

**Findings:**
- Cohorts **Kings, Quincy, and Joshua** all look at PHP in their top 10 most visited lessons. They also have differently marked pathways to their content which could suggest that they made significant changes before or after these cohorts.
- Regression lesson is most viewed for **Curie** with about 600 clicks where as it barely shows up one other time in the top 6 lessons for other data science cohorts

In [8]:
#make df of just those in Data Science 
DS = df[df.program_id == 'Data Science']

#make df of just those in WebDev 
WebDev = df[(df.program_id == 'PHP') | (df.program_id == 'Java') | (df.program_id == 'Front End') ]

In [12]:
DS_Cohorts = DS.groupby('page')['name'].count().sort_values()

In [15]:
DS_Cohorts

page
%20https://github.com/RaulCPena                      1
appendix/open_data/www.census.gov                    1
appendix/open_data/www.flickr.com/services/api       1
appendix/open_data/www.openstreetmap.org             1
appendix/univariate_regression_in_excel              1
                                                  ... 
1-fundamentals/AI-ML-DL-timeline.jpg              1651
1-fundamentals/modern-data-scientist.jpg          1655
classification/overview                           1785
search/search_index.json                          2203
/                                                 8358
Name: name, Length: 682, dtype: int64

In [13]:
WebDev_Cohorts = WebDev.groupby('page')['name'].count().sort_values()

In [14]:
WebDev_Cohorts

page
'                               1
cli-02-listing-files            1
cli-03-file-paths               1
cli-07-more-topics              1
codeup-exercises                1
                            ...  
java-iii                    13162
search/search_index.json    15331
toc                         17580
javascript-i                18193
/                           37496
Name: name, Length: 2054, dtype: int64

In [17]:
WebDevDict = {}
for x in WebDev.name.unique():
    WebDevDict[x] = pd.DataFrame(df[df.name == x])

In [18]:
DSDict = {}
for x in DS.name.unique():
    DSDict[x] = pd.DataFrame(df[df.name == x])

In [45]:
#cohorts Kings, quincy, and joshua all look at PHP in their top 10 most visited lessons. they also have differently marked pathways to their content which could suggest that they made significant changes before or after these cohorts.
for x in WebDevDict:
    display(x, WebDevDict[x].groupby('page')['name'].count().sort_values(ascending=False).head(10))


'Hampton'

page
java-iii/servlets     30
mysql                 30
javascript-ii         31
javascript-i          36
spring                43
java-i                46
java-ii               46
appendix              55
java-iii              57
/                    210
Name: name, dtype: int64

'Teddy'

page
appendix                     505
java-ii                      509
jquery                       512
java-i                       522
mkdocs/search_index.json     595
javascript-i                 615
mysql                        638
java-iii                     732
spring                       744
/                           1828
Name: name, dtype: int64

'Sequoia'

page
spring/fundamentals/form-model-binding    156
mkdocs/search_index.json                  168
html-css                                  168
appendix                                  170
spring/fundamentals/repositories          172
spring/fundamentals/controllers           190
javascript-i                              202
spring/fundamentals/views                 207
spring                                    364
/                                         630
Name: name, dtype: int64

'Arches'

page
java-i           132
jquery           133
appendix         133
java-iii         153
java-ii          167
spring           192
javascript-ii    204
html-css         215
javascript-i     294
/                626
Name: name, dtype: int64

'Niagara'

page
java-ii          17
java-i           18
appendix         20
javascript-ii    21
javascript-i     23
mysql            24
java-iii         29
/                37
index.html       63
spring           65
Name: name, dtype: int64

'Pinnacles'

page
mysql                 39
java-iii              45
java-ii               50
javascript-ii/npm     51
java-i                54
spring                56
html-css              56
javascript-ii         58
javascript-i          69
/                    149
Name: name, dtype: int64

'Quincy'

page
content/appendix                                     14
content/php_iii/php-with-html                        15
javascript-i/conditionals                            16
content/php_iii                                      19
content/mysql/intro-to-mysql                         19
content/laravel/intro/application-structure.html     20
content/laravel/intro/environment.html               31
content/html-css                                     33
content/laravel/intro                                38
/                                                   151
Name: name, dtype: int64

'Kings'

page
content/laravel/quickstart/db-migrations.html        40
content/php_ii                                       40
content/php_iv                                       42
content/html-css                                     45
content/php_iii                                      47
content/laravel/intro/gitbook/images/favicon.ico     56
content/laravel/intro/application-structure.html     63
content/laravel/intro                                83
index.html                                           84
/                                                   219
Name: name, dtype: int64

'Lassen'

page
spring/fundamentals/repositories    115
/                                   125
java-i                              138
java-ii                             160
appendix                            170
html-css                            174
spring                              222
java-iii                            224
javascript-i                        233
index.html                          877
Name: name, dtype: int64

'Mammoth'

page
content/php_i/intro-via-interactive-shell     9
appendix                                      9
javascript-ii                                10
java-ii                                      10
spring/fundamentals/views                    11
spring                                       12
spring/fundamentals/controllers              12
java-iii                                     13
java-i                                       19
/                                            33
Name: name, dtype: int64

'Glacier'

page
javascript-i/javascript-with-html                                            14
javascript-i/conditionals                                                    16
javascript-i/introduction/operators                                          18
javascript-i/introduction/variables                                          19
javascript-i/functions                                                       20
content/html-css/introduction.html                                           20
html-css                                                                     21
javascript-i/introduction/working-with-data-types-operators-and-variables    24
content/html-css                                                             29
/                                                                            51
Name: name, dtype: int64

'Denali'

page
/                            1
mkdocs/search_index.json     1
prework/databases            1
prework/versioning/github    1
Name: name, dtype: int64

'Joshua'

page
content/javascript/primitive-types.html     7
content/php_iv                              7
content/html-css/forms.html                 7
index.html                                  8
/                                           8
content/html-css/elements.html              9
content/mysql/intro-to-mysql               10
content/javascript                         12
content/html-css/introduction.html         14
content/html-css                           19
Name: name, dtype: int64

'Olympic'

page
javascript-i/bom-and-dom/bom                                                  60
java-i/console-io                                                             60
javascript-i/javascript-with-html                                             63
javascript-i/introduction/working-with-data-types-operators-and-variables     65
html-css                                                                      69
java-i/methods                                                                69
jquery                                                                        71
java-i                                                                        76
javascript-i                                                                 128
/                                                                            249
Name: name, dtype: int64

'Ulysses'

page
jquery                       364
mysql                        367
java-i                       377
spring                       409
java-iii                     412
java-ii                      425
javascript-i                 526
html-css                     562
mkdocs/search_index.json     721
/                           1641
Name: name, dtype: int64

'Badlands'

page
content/php_i/intro-via-interactive-shell/hello-world-and-other-strings.html     2
content/php_i/intro-to-operators                                                 2
content/laravel/intro                                                            2
content/php_i/intro-via-interactive-shell                                        3
content/php_ii                                                                   4
content/php_ii/command-line/io-streams.html                                      5
content/php_i                                                                    5
content/php_ii/control-structures-i                                              5
content/php_ii/command-line                                                      6
/                                                                               17
Name: name, dtype: int64

'Apollo'

page
/                                              1
content/html-css/gitbook/images/favicon.ico    1
content/html-css/introduction.html             1
content/html-css                               2
Name: name, dtype: int64

'Ike'

page
toc                       6
index.html                6
html-css/css-i            6
javascript-i              7
html-css/forms            8
content/html-css          9
html-css/introduction    10
html-css/elements        11
html-css                 12
/                        19
Name: name, dtype: int64

'Voyageurs'

page
html-css          528
jquery            583
javascript-ii     584
java-i            641
spring            650
mysql             663
java-ii           756
java-iii          770
javascript-i      884
/                2101
Name: name, dtype: int64

'Wrangell'

page
html-css/css-ii/bootstrap-grid-system     313
java-ii                                   325
java-iii                                  337
jquery                                    377
java-i                                    406
html-css                                  411
search/search_index.json                  504
javascript-i                              560
toc                                       991
/                                        1132
Name: name, dtype: int64

'Xanadu'

page
java-i                      484
javascript-ii               496
java-iii                    497
java-ii                     502
mysql                       509
jquery                      544
search/search_index.json    577
html-css                    597
javascript-i                722
/                           924
Name: name, dtype: int64

'Franklin'

page
spring/setup                    2
javascript-ii                   2
javascript-i/coffee-project     2
javascript-i                    2
java-iii/user-input             3
java-iii                        3
java-ii                         3
java-i                          3
/                               4
javascript-ii/es6              10
Name: name, dtype: int64

'Yosemite'

page
javascript-ii/map-filter-reduce    299
mysql                              331
javascript-ii                      337
java-iii                           351
search/search_index.json           361
spring                             381
html-css                           384
javascript-i                       452
toc                                697
/                                  981
Name: name, dtype: int64

'Staff'

page
javascript-ii                993
jquery/ajax/weather-map     1050
java-ii                     1260
html-css                    1284
appendix                    1288
search/search_index.json    1349
java-iii                    1393
spring                      1403
javascript-i                1817
/                           6340
Name: name, dtype: int64

'Zion'

page
java-i                       605
java-ii                      624
javascript-ii                647
spring                       672
html-css                     675
search/search_index.json     700
java-iii                     753
javascript-i                 897
toc                         1465
/                           1798
Name: name, dtype: int64

'Andromeda'

page
java-i           338
jquery           338
mysql            355
html-css         375
java-ii          398
java-iii         433
spring           487
javascript-i     516
toc              638
/               1174
Name: name, dtype: int64

'Betelgeuse'

page
html-css                    457
java-iii                    472
java-ii                     481
java-i                      482
html-css/elements           495
jquery                      542
toc                         610
javascript-i                717
search/search_index.json    761
/                           955
Name: name, dtype: int64

'Ceres'

page
spring                       562
mysql                        622
jquery                       637
java-ii                      681
java-iii                     682
html-css                     766
toc                          911
javascript-i                1003
search/search_index.json    1380
/                           1653
Name: name, dtype: int64

'Deimos'

page
java-ii                      511
mysql                        517
java-iii                     523
spring                       533
jquery                       563
search/search_index.json     746
toc                          750
html-css                     773
javascript-i                 896
/                           1467
Name: name, dtype: int64

'Europa'

page
html-css                     353
jquery                       370
mysql                        372
java-ii                      375
html-css/elements            430
java-iii                     451
javascript-i                 506
search/search_index.json     742
toc                          964
/                           1295
Name: name, dtype: int64

'Fortuna'

page
java-i                       555
html-css                     585
mysql                        591
spring                       636
java-ii                      657
javascript-i                 785
java-iii                     786
search/search_index.json    1020
toc                         1293
/                           2038
Name: name, dtype: int64

'Apex'

page
mysql                        483
java-i                       590
spring                       600
java-iii                     663
java-ii                      666
javascript-i                 669
html-css                     738
toc                          810
/                           1346
search/search_index.json    1497
Name: name, dtype: int64

'Ganymede'

page
java-i                       515
javascript-ii                525
jquery                       536
appendix                     561
java-ii                      641
java-iii                     651
javascript-i                 760
toc                          822
search/search_index.json    1095
/                           1727
Name: name, dtype: int64

'Everglades'

page
/    1
Name: name, dtype: int64

'Hyperion'

page
java-i                       489
jquery                       497
spring                       526
mysql                        553
search/search_index.json     670
java-ii                      688
java-iii                     701
javascript-i                 910
toc                         1022
/                           1365
Name: name, dtype: int64

'Bash'

page
javascript-i/introduction    256
jquery                       262
html-css                     265
java-ii                      268
spring                       311
java-iii                     324
javascript-i                 374
toc                          404
search/search_index.json     660
/                            772
Name: name, dtype: int64

'Jupiter'

page
java-i                       503
mysql                        564
spring                       621
java-ii                      755
html-css                     784
java-iii                     795
javascript-i                 926
search/search_index.json     998
/                           1696
toc                         1866
Name: name, dtype: int64

'Kalypso'

page
java-i                      408
java-ii                     417
javascript-ii               440
java-iii                    456
search/search_index.json    467
jquery                      470
html-css                    495
toc                         495
javascript-i                631
/                           939
Name: name, dtype: int64

'Luna'

page
javascript-ii                                                                247
java-ii                                                                      255
java-iii                                                                     259
jquery                                                                       301
html-css                                                                     307
javascript-i/introduction/working-with-data-types-operators-and-variables    315
toc                                                                          373
search/search_index.json                                                     405
/                                                                            491
javascript-i                                                                 676
Name: name, dtype: int64

'Marco'

page
javascript-i/conditionals                                                    292
javascript-i/loops                                                           312
javascript-i/functions                                                       382
jquery                                                                       405
javascript-i/javascript-with-html                                            411
html-css                                                                     479
javascript-i/introduction/working-with-data-types-operators-and-variables    505
/                                                                            699
javascript-i                                                                 907
toc                                                                          910
Name: name, dtype: int64

'Neptune'

page
javascript-i/functions                                                       217
javascript-i/loops                                                           226
html-css/css-ii/bootstrap-introduction                                       227
html-css/css-ii/bootstrap-grid-system                                        256
javascript-i                                                                 272
javascript-i/javascript-with-html                                            311
html-css                                                                     313
javascript-i/introduction/working-with-data-types-operators-and-variables    329
toc                                                                          451
/                                                                            472
Name: name, dtype: int64

'Oberon'

page
javascript-i/introduction/variables                                           95
toc                                                                          101
javascript-i/javascript-with-html                                            125
/                                                                            131
javascript-i                                                                 136
javascript-i/conditionals                                                    144
javascript-i/introduction/working-with-data-types-operators-and-variables    146
javascript-i/functions                                                       171
javascript-i/introduction/primitive-types                                    177
javascript-i/introduction/operators                                          179
Name: name, dtype: int64

In [46]:
# Regression lesson in most viewed for Curie  with ab 600 clicks where as it barely shows up one other time in the top 6 lessons for other data science cohorts
for x in DSDict:
    display(x, DSDict[x].groupby('page')['name'].count().sort_values(ascending=False).head(7))


'Bayes'

page
10-anomaly-detection/AnomalyDetectionCartoon.jpeg     387
6-regression/1-overview                               521
search/search_index.json                              588
1-fundamentals/1.1-intro-to-data-science              640
1-fundamentals/AI-ML-DL-timeline.jpg                  648
1-fundamentals/modern-data-scientist.jpg              650
/                                                    1967
Name: name, dtype: int64

'Curie'

page
3-sql/1-mysql-overview                       441
1-fundamentals/1.1-intro-to-data-science     461
1-fundamentals/AI-ML-DL-timeline.jpg         465
1-fundamentals/modern-data-scientist.jpg     467
search/search_index.json                     538
6-regression/1-overview                      595
/                                           1712
Name: name, dtype: int64

'Darden'

page
anomaly-detection/overview                         592
anomaly-detection/AnomalyDetectionCartoon.jpeg     612
search/search_index.json                           664
sql/mysql-overview                                 774
classification/scale_features_or_not.svg           943
classification/overview                           1109
/                                                 2980
Name: name, dtype: int64

'Easley'

page
fundamentals/intro-to-data-science                                372
fundamentals/modern-data-scientist.jpg                            379
fundamentals/AI-ML-DL-timeline.jpg                                381
classification/classical_programming_vs_machine_learning.jpeg     432
classification/overview                                           445
classification/scale_features_or_not.svg                          463
/                                                                1115
Name: name, dtype: int64

'Florence'

page
sql/mysql-overview                        203
fundamentals/DataToAction_v2.jpg          208
python/data-types-and-variables           258
/                                         584
fundamentals/intro-to-data-science        615
fundamentals/AI-ML-DL-timeline.jpg        624
fundamentals/modern-data-scientist.jpg    627
Name: name, dtype: int64

## 3. Are there students who, when active, hardly access the curriculum? If so, what information do you have about these students?

**Findings:**
- There are 233 users that didn't access the curriculum during their cohort timeframe, but some of those don't have a cohort listed (78)
- I defined "hardly active" as those students who accessed less than or equal to 10% of the average usage (pages viewed)
- 54 users accessed less than or equal to 10% of the average number of pages viewed per user.
- All but one user was from the Java program, user_id 679 was data science(assumed quit the course)
- **Oberon** and **Sequoia** cohorts only had students access the curriculum up through the first month of the course. ***This might mean that these cohorts used a different curriculum platform or there is something wrong with the start and end dates in the db***
- user_ids 24, 41, 371, 741, 56, 57, 45, 36, 250, 65, 52, 33, 253, 278 only started accessing the curriculum when the cohort was almost finished 
- user_ids 879, 832,  388, 572, 772, 529, 543, 46, 13, 702, 806 accessed the curriculum on the first day (or first week) and then stopped. ***I assume these are students who dropped out.***



Below is the code that helped me discover these finidings.

In [7]:
#only look at students
students = df[df['name'] != 'Staff']

In [8]:
#find those who accessed the curriculum during their class
active_student_access = students.query('start_date < date < end_date')

In [13]:
usage = active_student_access.user_id.value_counts().sort_values()

In [14]:
#I defined "hardly active" as those students who accessed less than or equal to 10% of the average usage (pages viewed)
usage.mean() * .10

89.00359116022099

In [15]:
#how many accessed less than 10% of the average -- 54
usage[usage <= 89]

879     1
679     3
832     3
278     4
956     5
812     7
388     8
572    12
889    16
968    16
487    17
963    17
961    18
772    20
49     25
64     25
976    25
24     26
41     28
972    31
973    32
967    33
529    35
959    38
971    41
964    42
371    42
981    42
960    45
741    47
978    49
965    51
56     52
962    52
543    53
46     54
57     62
969    62
858    63
45     65
36     65
975    66
250    67
65     73
52     75
977    76
974    76
13     79
33     79
955    80
702    85
806    88
954    88
253    89
Name: user_id, dtype: int64

In [17]:
#test oddities of those who didn't view many times during cohort
active_student_access[active_student_access.user_id == 24]

date      time                                    page  \
date                                                                      
2018-01-26 2018-01-26  11:24:17                                appendix   
2018-01-26 2018-01-26  11:24:29              appendix/capstone-workbook   
2018-01-29 2018-01-29  11:25:49                                       /   
2018-01-29 2018-01-29  11:25:54                                  spring   
2018-01-29 2018-01-29  11:25:59               spring/fundamentals/views   
2018-01-29 2018-01-29  13:28:23                mkdocs/search_index.json   
2018-01-29 2018-01-29  13:28:33                                html-css   
2018-01-29 2018-01-29  13:28:41  html-css/css-ii/bootstrap-introduction   
2018-01-29 2018-01-29  13:29:50                                       /   
2018-01-29 2018-01-29  13:29:53                                html-css   
2018-01-29 2018-01-29  13:29:56  html-css/css-ii/bootstrap-introduction   
2018-01-30 2018-01-30  09:30:44                            javascript-i   
2018-01-30 2018-01-30  09:30:49            javascript-i/google-maps-api   
2018-01-30 2018-01-30  09:32:13                mkdocs/search_index.json   
2018-01-30 2018-01-30  09:32:21                mkdocs/search_index.json   
2018-02-01 2018-02-01  10:21:40                                  spring   
2018-02-01 2018-02-01  10:21:45         spring/fundamentals/controllers   
2018-02-01 2018-02-01  10:22:10                                java-iii   
2018-02-01 2018-02-01  10:22:16    java-iii/deployment-and-dependencies   
2018-02-01 2018-02-01  10:22:38      slides/deployment_and_dependencies   
2018-02-01 2018-02-01  10:22:46      slides/deployment_and_dependencies   
2018-02-01 2018-02-01  10:22:51                       java-iii/servlets   
2018-02-01 2018-02-01  10:22:56                         slides/servlets   
2018-02-01 2018-02-01  10:22:59                         slides/servlets   
2018-02-07 2018-02-07  11:07:28                                appendix   
2018-02-07 2018-02-07  11:07:31              appendix/capstone-workbook   

            user_id  cohort_id            ip     name start_date   end_date  \
date                                                                          
2018-01-26       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-26       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-29       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-30       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-30       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-30       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-01-30       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24       21.0  97.105.19.61  Sequoia 2017-09-27 2018-02-15   
2018-02-01       24 

## 4. Is there any suspicious activity, such as users/machines/etc accessing the curriculum who shouldn’t be? Does it appear that any web-scraping is happening? Are there any suspicious IP addresses?

## 5. At some point in 2019, the ability for students and alumni to access both curriculums (web dev to ds, ds to web dev) should have been shut off. Do you see any evidence of that happening? Did it happen before?

## 6. What topics are grads continuing to reference after graduation and into their jobs (for each program)?

<div class="alert alert-info">

    
# Q#6 
# ANSWER:

## 1. Data Science
    
|Curriculum        |Click Count|Student Count|
|:-----------------|:-:|:-:|
|sql/mysql-overview|275|38|
|classification/overview|266|39|
|anomaly-detection/overview|191|31|
|fundamentals/intro-to-data-science|184|36|

## 2. PHP

|Curriculum        |Click Count|Student Count|
|:-----------------|:-:|:-:|
|javascript-i	|736|	35|
|html-css	|542|	40|
|spring	|501|	25|
|java-iii	|479|	20|
    
## 3. Java
    
|Curriculum        |Click Count|Student Count|
|:-----------------|:-:|:-:|    
|javascript-i	|4229|	300|
|spring	|3760|	306|
|html-css	|3136|	296|
|java-iii	|3058|	273|
    
## 4. Front End    
    
|Curriculum        |Click Count|Student Count|
|:-----------------|:-:|:-:|    
|content/html-css	|2	|1|
|content/html-css/introduction.html	|1	|1|    

# Grad Users

In [6]:
# Create DataFrame for users that have accessed the curriculum past there graduation date
grad_users = df[df.date > df.end_date]

In [7]:
grad_users.T

date                 2018-01-26           2018-01-26  \
date        2018-01-26 00:00:00  2018-01-26 00:00:00   
time                   09:55:03             09:56:02   
page                          /              java-ii   
user_id                       1                    1   
cohort_id                   8.0                  8.0   
ip                 97.105.19.61         97.105.19.61   
name                    Hampton              Hampton   
start_date  2015-09-22 00:00:00  2015-09-22 00:00:00   
end_date    2016-02-06 00:00:00  2016-02-06 00:00:00   
program_id                  PHP                  PHP   

date                                 2018-01-26  \
date                        2018-01-26 00:00:00   
time                                   09:56:05   
page        java-ii/object-oriented-programming   
user_id                                       1   
cohort_id                                   8.0   
ip                                 97.105.19.61   
name                                    Hampton   
start_date                  2015-09-22 00:00:00   
end_date                    2016-02-06 00:00:00   
program_id                                  PHP   

date                                2018-01-26           2018-01-26  \
date                       2018-01-26 00:00:00  2018-01-26 00:00:00   
time                                  09:56:06             10:14:47   
page        slides/object_oriented_programming                    /   
user_id                                      1                   11   
cohort_id                                  8.0                  1.0   
ip                                97.105.19.61         97.105.19.61   
name                                   Hampton               Arches   
start_date                 2015-09-22 00:00:00  2014-02-04 00:00:00   
end_date                   2016-02-06 00:00:00  2014-04-22 00:00:00   
program_id                                 PHP                  PHP   

date                 2018-01-26              2018-01-26  \
date        2018-01-26 00:00:00     2018-01-26 00:00:00   
time                   10:14:51                10:15:02   
page               javascript-i  javascript-i/functions   
user_id                      11                      11   
cohort_id                   1.0                     1.0   
ip                 97.105.19.61            97.105.19.61   
name                     Arches                  Arches   
start_date  2014-02-04 00:00:00     2014-02-04 00:00:00   
end_date    2014-04-22 00:00:00     2014-04-22 00:00:00   
program_id                  PHP                     PHP   

date                    2018-01-26           2018-01-26           2018-01-26  \
date           2018-01-26 00:00:00  2018-01-26 00:00:00  2018-01-26 00:00:00   
time                      10:40:15             10:59:02             10:59:06   
page        javascript-i/functions           index.html               java-i   
user_id                          1                   21                   21   
cohort_id                      8.0                 16.0                 16.0   
ip                    97.105.19.61        71.137.146.71        71.137.146.71   
name                       Hampton              Niagara              Niagara   
start_date     2015-09-22 00:00:00  2016-10-26 00:00:00  2016-10-26 00:00:00   
end_date       2016-02-06 00:00:00  2017-03-09 00:00:00  2017-03-09 00:00:00   
program_id                     PHP                 Java                 Java   

date        ...                 2021-04-21           2021-04-21  \
date        ...        2021-04-21 00:00:00  2021-04-21 00:00:00   
time        ...                   16:07:17             16:30:30   
page        ...  javascript-ii/RESTful-api             appendix   
user_id     ...                         64                  771   
cohort_id   ...                       28.0                 62.0   
ip          ...              71.150.217.33      172.125.226.175   
name        ...                      St

In [13]:
# LOOK AT GRAD USERS
grad_program_views = grad_users.groupby(['program_id'])['page'].agg(['count', 'nunique'])
grad_program_views.sort_values(by='count', ascending=False)

count  nunique
program_id                   
Java          157786     1656
PHP            30548      710
Data Science   11495      437
Front End          5        4

In [14]:
# LOOK AT WHOLE DF
df_program_views = df.groupby(['program_id'])['page'].agg(['count', 'nunique'])
df_program_views.sort_values(by='count', ascending=False)

count  nunique
program_id                   
Java          713365     1913
Data Science  103411      682
nan            52893     1112
PHP            30548      710
Front End          5        4

In [15]:
grad_users.groupby('program_id')['page'].value_counts()

program_id    page                                    
Data Science  /                                           1436
              search/search_index.json                     493
              sql/mysql-overview                           275
              classification/overview                      266
              classification/scale_features_or_not.svg     219
                                                          ... 
PHP           students                                       1
              students/468/notes                             1
              students/units/75/sub_units/268                1
              teams/13                                       1
              uploads/58a217a705bde.jpg                      1
Name: page, Length: 2807, dtype: int64

# Data Science

In [16]:
# Create DataFrame for Data Science users that have accessed the curriculum past there graduation date
ds_grad_users = grad_users[grad_users.program_id == 'Data Science']

In [18]:
# Count how many times each page was accessed
ds_grad_users.page.value_counts().head(10)

/                                                 1436
search/search_index.json                           493
sql/mysql-overview                                 275
classification/overview                            266
classification/scale_features_or_not.svg           219
anomaly-detection/AnomalyDetectionCartoon.jpeg     193
anomaly-detection/overview                         191
fundamentals/AI-ML-DL-timeline.jpg                 189
fundamentals/modern-data-scientist.jpg             187
fundamentals/intro-to-data-science                 184
Name: page, dtype: int64

In [19]:
# Look at curriculum by click count and student nunique
ds_program_views = ds_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
ds_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                                          
/                                                1436       57
search/search_index.json                          493       42
sql/mysql-overview                                275       38
classification/overview                           266       39
classification/scale_features_or_not.svg          219       36
anomaly-detection/AnomalyDetectionCartoon.jpeg    193       31
anomaly-detection/overview                        191       31
fundamentals/AI-ML-DL-timeline.jpg                189       36
fundamentals/modern-data-scientist.jpg            187       35
fundamentals/intro-to-data-science                184       36

## Check the min/max Graduation Dates.

In [20]:
# Create Java for DS users
DS_users = df[df.program_id == 'Data Science']

In [21]:
# show max end date for DS users
DS_users.end_date.min()

Timestamp('2020-01-30 00:00:00')

In [22]:
# show max end date for DS users
DS_users.end_date.max()

Timestamp('2021-09-03 00:00:00')

In [23]:
# show min access date for df
df.date.min()

Timestamp('2018-01-26 00:00:00')

# PHP

In [24]:
# Create DataFrame for PHP users that have accessed the curriculum past there graduation date
php_grad_users = grad_users[grad_users.program_id == 'PHP']

In [27]:
# Look at curriculum by click count and student nunique
php_program_views = php_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
php_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                         
/               1681       90
index.html      1011       20
javascript-i     736       35
html-css         542       40
spring           501       25
java-iii         479       20
java-ii          454       24
java-i           444       40
javascript-ii    429       33
appendix         409       25

## Why are Curriculum access counts the same for PHP during and after graduation?

In [28]:
# Create DataFrame for PHP users
php_users = df[df.program_id == 'PHP']

In [29]:
# show max end date for Front End users
php_users.end_date.max()

Timestamp('2017-09-22 00:00:00')

In [30]:
# show min access date for df
df.date.min()

Timestamp('2018-01-26 00:00:00')

In [31]:
# Count how many times each page was accessed
php_users.page.value_counts().head(10)

/                1681
index.html       1011
javascript-i      736
html-css          542
spring            501
java-iii          479
java-ii           454
java-i            444
javascript-ii     429
appendix          409
Name: page, dtype: int64

## ANSWER: Because the last of the PHP users graduated in 2017, before the start date of our data access collection which begins in 2018.

# JAVA

In [32]:
# Create DataFrame for Java users that have accessed the curriculum past there graduation date
java_grad_users = grad_users[grad_users.program_id == 'Java']

In [34]:
# Count how many times each page was accessed
java_grad_users.page.value_counts().head(10)

/                           12406
javascript-i                 4229
spring                       3760
search/search_index.json     3562
html-css                     3136
java-iii                     3058
java-ii                      2985
java-i                       2679
appendix                     2662
javascript-ii                2549
Name: page, dtype: int64

In [35]:
# Look at curriculum by click count and student nunique
java_program_views = java_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
java_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                    
/                         12406      406
javascript-i               4229      300
spring                     3760      306
search/search_index.json   3562      262
html-css                   3136      296
java-iii                   3058      273
java-ii                    2985      274
java-i                     2679      300
appendix                   2662      279
javascript-ii              2549      264

## Check the min/max Graduation Dates.

In [36]:
# Create Java for PHP users
java_users = df[df.program_id == 'Java']

In [37]:
# show max end date for Java users
java_users.end_date.min()

Timestamp('2014-02-04 00:00:00')

In [38]:
# show max end date for Java users
java_users.end_date.max()

Timestamp('2021-10-01 00:00:00')

In [39]:
# show min access date for df
df.date.min()

Timestamp('2018-01-26 00:00:00')

# Front End

In [40]:
# Create DataFrame for Front End users that have accessed the curriculum past there graduation date
fe_grad_users = grad_users[grad_users.program_id == 'Front End']

In [42]:
# Count how many times each page was accessed
fe_grad_users.page.value_counts().head(10)

content/html-css                               2
/                                              1
content/html-css/gitbook/images/favicon.ico    1
content/html-css/introduction.html             1
Name: page, dtype: int64

In [43]:
# Look at curriculum by click count and student nunique
fe_program_views = fe_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
fe_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                                       
content/html-css                                 2        1
/                                                1        1
content/html-css/gitbook/images/favicon.ico      1        1
content/html-css/introduction.html               1        1

In [48]:
# Look at curriculum by click count and student nunique
ds_program_views = ds_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
ds_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                                          
/                                                1436       57
search/search_index.json                          493       42
sql/mysql-overview                                275       38
classification/overview                           266       39
classification/scale_features_or_not.svg          219       36
anomaly-detection/AnomalyDetectionCartoon.jpeg    193       31
anomaly-detection/overview                        191       31
fundamentals/AI-ML-DL-timeline.jpg                189       36
fundamentals/modern-data-scientist.jpg            187       35
fundamentals/intro-to-data-science                184       36

In [49]:
# Look at curriculum by click count and student nunique
php_program_views = php_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
php_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                         
/               1681       90
index.html      1011       20
javascript-i     736       35
html-css         542       40
spring           501       25
java-iii         479       20
java-ii          454       24
java-i           444       40
javascript-ii    429       33
appendix         409       25

In [50]:
# Look at curriculum by click count and student nunique
java_program_views = java_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
java_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                    
/                         12406      406
javascript-i               4229      300
spring                     3760      306
search/search_index.json   3562      262
html-css                   3136      296
java-iii                   3058      273
java-ii                    2985      274
java-i                     2679      300
appendix                   2662      279
javascript-ii              2549      264

In [51]:
# Look at curriculum by click count and student nunique
fe_program_views = fe_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
fe_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                                       
content/html-css                                 2        1
/                                                1        1
content/html-css/gitbook/images/favicon.ico      1        1
content/html-css/introduction.html               1        1

## 7. Which lessons are least accessed?

**Findings:**
- Data Science accessed pages with the prefix appendix the least.
- Front End accessed pages with the prefix content/html-css the least.
- Java accessed pages with the prefix cli- the least.
- PHP accessed pages with the prefix content/appendix/javascript/functions/ the least.

In [20]:
#Look at what page was least viewed for data science
ds.groupby(['page'])['page'].count().sort_values()

page
%20https://github.com/RaulCPena                      1
appendix/open_data/www.census.gov                    1
appendix/open_data/www.flickr.com/services/api       1
appendix/open_data/www.openstreetmap.org             1
appendix/univariate_regression_in_excel              1
                                                  ... 
1-fundamentals/AI-ML-DL-timeline.jpg              1651
1-fundamentals/modern-data-scientist.jpg          1655
classification/overview                           1785
search/search_index.json                          2203
/                                                 8358
Name: page, Length: 682, dtype: int64

In [21]:
#Look at what page was least viewed for front end
fe.groupby(['page'])['page'].count().sort_values()

page
/                                              1
content/html-css/gitbook/images/favicon.ico    1
content/html-css/introduction.html             1
content/html-css                               2
Name: page, dtype: int64

In [22]:
#Look at what page was least viewed for java
jv.groupby(['page'])['page'].count().sort_values()

page
'                                            1
cheatsheet.tex                               1
classification/user-defined-functions        1
cli-02-listing-files                         1
cli-03-file-paths                            1
                                         ...  
java-iii                                 12683
search/search_index.json                 15212
toc                                      17428
javascript-i                             17457
/                                        35814
Name: page, Length: 1913, dtype: int64

In [23]:
#Look at what page was least viewed for php
php.groupby(['page'])['page'].count().sort_values()

page
html-css/media-queries                                       1
prework/cli/07-editing-files-with-vim                        1
content/appendix/javascript/functions/controllers.html       1
content/appendix/javascript/functions/models.html            1
content/appendix/javascript/functions/scope.html             1
                                                          ... 
spring                                                     501
html-css                                                   542
javascript-i                                               736
index.html                                                1011
/                                                         1681
Name: page, Length: 710, dtype: int64

## 8. Anything else I should be aware of?

### Why are there so few Front End Graduates accessing the Curriculum?

In [44]:
# Create DataFrame for Front End users
fe_users = df[df.program_id == 'Front End']

In [45]:
# show max end date for Front End users
fe_users.end_date.max()

Timestamp('2015-07-29 00:00:00')

In [46]:
# show min access date for df
df.date.min()

Timestamp('2018-01-26 00:00:00')

In [47]:
# How many users and cohorts make up the Front End
fe_users.groupby(['name'])['user_id'].agg(['nunique'])

nunique
name           
Apollo        1

<div class="alert alert-info">


## ANSWER: Because there was only one Front End user who graduated in 2015, three years before the start date of our data access collection which begins in 2018.

## Most likely, the majority of the Null entries for Cohort could probably be attributed to Front End as well as a few towards Staff